packages/fcl/src/wallet-provider-spec/draft-v3.md

```diff
@@ -0,0 +1,548 @@
+
+In the following paragraphs we'll explore ways in which you, as a wallet developer, can integrate with FCL by providing implementataions of various FCL services. 
+
+The following services will be covered:
+- Authentication (Authn) Service
+- Authorization (Authz) Service
+- User Signature Service
+- Pre-Authz Service
+
+# Authentication
+
+In the following examples, we'll walk you through the process of building an authentication process.
+
+In FCL, wallets are configured by passing in a wallet provider's authentication URL as the `discovery.wallet` config variable.
+
+As someone who is making an FCL compatible wallet, you will need to make and expose a webpage or API hosted at an authentication endpoint that FCL will use.
+
+```javascript
+// IN APPLICATION
+// configuring fcl to point at a wallet looks like this
+import {config} from "@onflow/fcl"
+
+config({
+  "discovery.wallet": "your-url-that-fcl-will-use-for-authentication",
+  "discovery.wallet.method": "IFRAME/RPC" // Available methods are "IFRAME/RPC", "POP/RPC", "TAB/RPC" or "HTTP/POST"
+})
+```
+
+If the method specified is `IFRAME/RPC`, `POP/RPC` or `TAB/RPC`, then the URL specified as `discovery.wallet` will be rendered as a webpage. Otherwise, if the method specified is `HTTP/POST`, then the authentication process will happen over HTTP requests.
+
+Once the Authentication webpage is rendered, or the API is ready, the wallet then needs to tell FCL that it is ready. You will do this by sending a message to FCL, and FCL will send back a message with some additional information that you can use about the application requesting authentication on behalf of the user.
+
+```javascript
+// IN WALLET AUTHENTICATION FRAME
+import {WalletUtils} from "@onflow/fcl"
+
+function callback({ data }) {
+  if (typeof data != "object") return
+  if (typeof data.type !== "FCL:FRAME:READY:RESPONSE") return
+
+  ... // Do authentication things ...
+
+  WalletUtils.sendMsgToFCL("PollingResponse", {
+    "f_vsn": "1.0.0",
+    "status": "APPROVED", // PENDING | APPROVED | DECLINED
+    "data": {
+         ...
+    }
+  })
+}
+// add event listener first
+window.onMsgFromFCL("FCL:FRAME:READY:RESPONSE", callback)
+
+// tell fcl the wallet is ready
+WalletUtils.sendMsgToFCL("FCL:VIEW:READY")
+```
+
+During authentication, the application has a chance to request to you, the wallet, what they would like you to send back to them. These requests are included in the `FCL:FRAME:READY:RESPONSE` messsage sent to the wallet from FCL.
+
+An example of such a request is the OpenID service. The application can request for you to send them the email address of the current user. The application requesting this information does not mean you need to send it. It's entirely optional for you to do so.
```

Comment:
We might note that failing to provide the requested scope info may cause the application to not work.